plot tests: ensure all figures are closed

yashsaboo opened the pull request:
Test `testxyincrease_false_changes_axes` is flaky in nature.

TestCase passes consistently in the actual order with command:
```
pytest tests/test_plot.py
```

TestCase fails when one randomises the order in *test_plot.py* module by using the plugin [pytest-random-order](https://pypi.org/project/pytest-random-order/) with command:
```
pytest tests/test_plot.py --random-order-seed=802425 -v
```

The victim `test_xyincrease_false_changes_axes` fails in a failing test order because there is at least one test that runs before the victim and “pollute” the state (e.g., global state: here, it is `matplotlib's plt`) on which the victim depends. [(Paper)](https://taoxie.cs.illinois.edu/publications/esecfse19-ifixflakies.pdf)

After investigation, we found the polluter test to be *test_ylim_kwarg*

Potential Fix: Adding a `plt.clf` or `plt.cla` at the end of the polluter test. Let us know your preference.

Please let us know if you require the pass and fail logs.

Also, if one runs test_plot.py in any random order, Teardown phase throws an error which is raised by test_all_figures_closed(). 
Should I raise a separate PR for the same?

Contributors:
[Yash Saboo](https://github.com/yashsaboo)
[Nirupam K N](https://github.com/Nirupamkn)

mathause commented:
Thanks for looking into this. Yes, closing the figure at the beginning is brittle, unless _every_ test does it. However, it cannot be done at the end because else the figure is not closed if there is an error in the test (and the figure is still open at the start of the next test).

My preference is to enclose the test in a `figure_context` as already done for some of the tests:
https://github.com/pydata/xarray/blob/9533c92c675a30b8c9803da8d5c9bc529763f3eb/xarray/tests/test_plot.py#L2456-L2457

`figure_context` is a small context manager that wraps a `try: ... finally`  block, closing all open figures.
https://github.com/pydata/xarray/blob/9533c92c675a30b8c9803da8d5c9bc529763f3eb/xarray/tests/test_plot.py#L49

I think all `plt.clf()` instances should be replaced with `figure_context`. I missed those in my PR (#4365). Let us know if you are up to that.

cc @dcherian

yashsaboo commented:
Thanks a lot for the feedback.
We have addressed all the above issues and made the necessary changes. Please let us know your thoughts on these new changes.

mathause commented:
Thanks! There are two more instances that would be good to wrap in a `figure_context`.

https://github.com/pydata/xarray/blob/a28f0df559cf5a18fce6c4a78277d3ef60898c0f/xarray/tests/test_plot.py#L316

https://github.com/pydata/xarray/blob/a28f0df559cf5a18fce6c4a78277d3ef60898c0f/xarray/tests/test_plot.py#L1759

If you want you can also give yourselves credit in [whats-new](https://github.com/pydata/xarray/blob/master/doc/whats-new.rst).

yashsaboo commented:
Thanks a lot again for the feedback. Appreciate it! We have made the necessary changes.

mathause reviewed (commented):
Two more nits from my side.

yashsaboo commented on xarray/tests/test_plot.py:
```diff
@@ -313,12 +313,11 @@ def test_2d_coords_line_plot(self):
+        with figure_context():
+            hdl = da.plot.line(x="lon", hue="x")
+            assert len(hdl) == 5
+            hdl = da.plot.line(x="lon", hue="y")
```
Thanks for the suggestion! Just to double-check, you want to wrap the two `plot-assert` into two different `figure_context()`, right? 
```
with figure_context():  
            hdl = da.plot.line(x="lon", hue="x")  
            assert len(hdl) == 5  
with figure_context():  
            hdl = da.plot.line(x="lon", hue="y")  
            assert len(hdl) == 4  
```

mathause replied:
Yes, correctly, such that the first figure gets closed before the second is created. For this test it is not strictly necessary because it works anyways. But it's certainly cleaner.

yashsaboo commented on xarray/tests/test_plot.py:
```diff
@@ -1756,14 +1755,13 @@ def test_regression_rgb_imshow_dim_size_one(self):
+            da.plot.imshow(origin="upper")
+            assert plt.xlim()[0] < 0
+            assert plt.ylim()[1] < 0
+            da.plot.imshow(origin="lower")
```
Same as above, right?

keewis commented:
how thorough should this be? If I understand it correctly, there are still a few calls left, e.g. in `test_plot_transposed_nondim_coord` (hidden by `getattr` calls).

mathause commented:
Good catch! If you haven't lost patience, there are two more things that leave open figures: `test_plot_transposed_nondim_coord` and some of the tests in `TestDiscreteColorMap`.

1. To fix the first both of these lines need to be enclosed in a `figure_context` (each line separately):

https://github.com/pydata/xarray/blob/f1e2f9d510a63480f4dd7917083cd982ac4d451d/xarray/tests/test_plot.py#L2431-L2432

2. To fix the second you can add the following code 

```python
        yield
        # Remove all matplotlib figures
        plt.close("all")
```
_after_ the following line:

https://github.com/pydata/xarray/blob/f1e2f9d510a63480f4dd7917083cd982ac4d451d/xarray/tests/test_plot.py#L964

And you should give yourselves credit in what's new. That's now more than a one-liner ;-)

---
Just as an addendum. The figures in the `PlotTestCase` class are autoclosed by

https://github.com/pydata/xarray/blob/f1e2f9d510a63480f4dd7917083cd982ac4d451d/xarray/tests/test_plot.py#L128-L133

but only if no more than one figure is opened per method.

yashsaboo commented:
Thanks, @keewis for bringing that point up. We were aware of `test_plot_transposed_nondim_coord` polluter but were still investigating the test cases it was polluting. FYI, you can try this --random-order-seed=846619. Please feel free to give more suggestions. Highly appreciate it!
@mathause, we have implemented all of the above changes. Thanks a lot again! :)

yashsaboo commented:
I am not sure why the check was unsuccessful. It says that the action in CI was terminated. 
I opened a new PR for the new amends [here](https://github.com/pydata/xarray/pull/4609). I closed it since it would be taken care here itself. Can that be the cause? I reopened it later though.

mathause reviewed (approved):
This closes all figures at the end of the tests. I'll merge in a few days to see if others have comments.

I changed the title of the PR. Let me know if you object.

yashsaboo commented:
Sounds good @mathause. Thanks again! :)

max-sixty commented:
Thanks @yashsaboo ! Great to have you as a contributor!

Nirupamkn commented:
> This closes all figures at the end of the tests. I'll merge in a few days to see if others have comments.
> 
> I changed the title of the PR. Let me know if you object.

@mathause @keewis Thanks a lot for all your feedback! 
I and @yashsaboo enjoyed working on this! And it was our first open source contribution! Hope to contribute more!

max-sixty commented:
@Nirupamkn @yashsaboo do you want to add an entry to the whatsnew file?

Nirupamkn commented:
> @Nirupamkn @yashsaboo do you want to add an entry to the whatsnew file?

Yeah sure! @max-sixty 
Is there any specific procedure to do it? or Just have to write 2 lines writeup at the end?

max-sixty commented:
Yes exactly, under "Internal Changes"

yashsaboo commented:
Thanks @max-sixty!! We have updated the [whats-new](https://github.com/pydata/xarray/blob/master/doc/whats-new.rst) file too now.

mathause commented:
Ok. let's get this in before the release. Thanks again @Nirupamkn @yashsaboo, happy to have you as contributors!

yashsaboo commented:
It was our pleasure. Thanks @mathause for all the help! :)
